In [57]:
import pandas as pd
import json 
import os

In [58]:

def extract_audio_links(folder_path, output_file):
    """
    Extract audio links from JSON files in the specified folder
    and create a combined JSON with {filename: audio_link} format.
    """
    # Dictionary to store filename: audio link pairs
    audio_links = {}
    
    # Check if the folder exists
    if not os.path.exists(folder_path):
        print(f"Error: Folder '{folder_path}' not found.")
        return None
    
    # Iterate through all files in the folder
    for filename in os.listdir(folder_path):
        # Only process JSON files
        if filename.endswith('.json'):
            file_path = os.path.join(folder_path, filename)
            
            try:
                # Read and parse the JSON file
                with open(file_path, 'r', encoding='utf-8') as file:
                    try:
                        data = json.load(file)
                        
                        # Extract the audio link
                        if (
                            'data' in data and 
                            'body' in data['data'] and 
                            'URL' in data['data']['meta_data']
                        ):
                            audio_link = data['data']['meta_data']['URL']
                            
                            # Only add to the dictionary if there's a valid audio link
                            if audio_link and audio_link != "":
                                # Remove the .json extension from the filename
                                filename_without_extension = os.path.splitext(filename)[0]
                                audio_links[filename_without_extension] = audio_link
                                
                    except json.JSONDecodeError:
                        print(f"Error: Unable to parse JSON in file '{filename}'")
            except Exception as e:
                print(f"Error processing file '{filename}': {str(e)}")
    
    # Write the combined data to a new JSON file
    
    # output_file = f"{categ}_combined_audio_links.json"
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(audio_links, f, ensure_ascii=False, indent=4)
    
    print(f"Successfully created '{output_file}' with {len(audio_links)} all article links.")
    return audio_links, len(audio_links)

In [76]:
categ = "གོང་ས་མཆོག"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

first_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/གོང་ས་མཆོག_combined_all_articles_links.json' with 1857 all article links.


In [77]:
categ = "བོད།"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

second_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/བོད།_combined_all_articles_links.json' with 2664 all article links.


In [78]:
categ = "བཙན་བྱོལ།"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

third_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/བཙན་བྱོལ།_combined_all_articles_links.json' with 4411 all article links.


In [79]:
categ = "འཛམ་གླིང༌།"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

forth_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/འཛམ་གླིང༌།_combined_all_articles_links.json' with 3474 all article links.


In [80]:
categ = "རྒྱ་དཀར་ནག"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

fif_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/རྒྱ་དཀར་ནག_combined_all_articles_links.json' with 2133 all article links.


In [81]:
categ = "སྤྱི་ཚོགས།"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

six_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/སྤྱི་ཚོགས།_combined_all_articles_links.json' with 3886 all article links.


In [65]:
categ = "གསར་འགྱུར།"
folder_path = f"./data/{categ}/"
file_name = f"{categ}_combined_all_articles_links.json"
output_file = "./data/" + file_name

sev_links, file_with_audio = extract_audio_links(folder_path, output_file)

Successfully created './data/གསར་འགྱུར།_combined_all_articles_links.json' with 28747 all article links.


## check duplicates

In [82]:
first_list = []
for keys, links in first_links.items():
    first_list.append(links)

print("total links", len(first_list))
print("duplicates in own category", len(set(first_list)))

total links 1857
duplicates in own category 1857


In [83]:
second_list = []
for keys, links in second_links.items():
    second_list.append(links)


print("total links", len(second_list))
print("duplicates in own category", len(set(second_list)))

total links 2664
duplicates in own category 2664


In [84]:
third_list = []
for keys, links in third_links.items():
    third_list.append(links)

print("total links", len(third_list))
print("duplicates in own category", len(set(third_list)))

total links 4411
duplicates in own category 4405


In [85]:
forth_list = []
for keys, links in forth_links.items():
    forth_list.append(links)

print("total links", len(forth_list))
print("duplicates in own category", len(set(forth_list)))

total links 3474
duplicates in own category 3474


In [86]:
fi_list = []
for keys, links in fif_links.items():
    fi_list.append(links)

print("total links", len(fi_list))
print("duplicates in own category", len(set(fi_list)))

total links 2133
duplicates in own category 2133


In [87]:
six_list = []
for keys, links in six_links.items():
    six_list.append(links)

print("total links", len(six_list))
print("duplicates in own category", len(set(six_list)))

total links 3886
duplicates in own category 3886


In [88]:
seven_list = []
for keys, links in sev_links.items():
    seven_list.append(links)

print("total links", len(seven_list))
print("duplicates in own category", len(set(seven_list)))

total links 28747
duplicates in own category 28695


# finding duplicates in inter categories

### 1. checking 1st and 2nd category as it's already downloaded

In [89]:
total_links = len(second_list) + len(first_list)  
print(f"total links {total_links}")

all_links = set()
all_links.update(first_list)
all_links.update(second_list)

non_dup_links  = len(all_links)
print(f"final links after removing duplicates {non_dup_links}")
print(f"removed duplicates {total_links-non_dup_links} - \npercentage: {round((total_links-non_dup_links)/total_links, 2)*100}% duplicate values")


total links 4521
final links after removing duplicates 4510
removed duplicates 11 - 
percentage: 0.0% duplicate values


- here we can see that we have only 0% duplicate, so we will ignore it and push to the AWS for faster process

### 2. Duplicate checks were performed across all categories, excluding གསར་འགྱུར།. 
- This category(གསར་འགྱུར།) contains approximately 28k articles, exceeding the combined article count of all other categories.

In [91]:
total_links = len(forth_list) + len(third_list) + len(second_list) + len(first_list)  + len(fi_list)  + len(six_list) 
print(f"total links {total_links}")

all_links = set()
all_links.update(first_list)
all_links.update(second_list)
all_links.update(third_list)
all_links.update(forth_list)
all_links.update(fi_list)
all_links.update(six_list)

non_dup_links  = len(all_links)
print(f"final links after removing duplicates {non_dup_links}")
print(f"removed duplicates {total_links-non_dup_links} - \npercentage: {round((total_links-non_dup_links)/total_links, 2)*100}% duplicate values")
s

total links 18425
final links after removing duplicates 16386
removed duplicates 2039 - 
percentage: 11.0% duplicate values


- here we can see that around 11% are duplicates, and we need to perform remove duplicate. 

### Checking for all categories including གསར་འགྱུར།. 

In [75]:
total_links = len(forth_list) + len(third_list) + len(second_list) + len(first_list)  + len(fi_list)  + len(six_list) +len(seven_list)
print(f"total links {total_links}")

all_links = set()
all_links.update(first_list)
all_links.update(second_list)
all_links.update(third_list)
all_links.update(forth_list)
all_links.update(fi_list)
all_links.update(six_list)
all_links.update(seven_list)

non_dup_links  = len(all_links)
print(f"final links after removing duplicates {non_dup_links}")
print(f"removed duplicates {total_links-non_dup_links} - \npercentage: {round((total_links-non_dup_links)/total_links, 2)*100}% duplicate values")


total links 47172
final links after removing duplicates 43919
removed duplicates 3253 - 
percentage: 7.000000000000001% duplicate values


- Here we can see that 3253 - 2039 = 1214 new duplicate found.
- This implies that the majority of articles in གསར་འགྱུར། are independent of the other categories

## making all the data as one df so we can filter it

- here we know that བོད། and གོང་ས་མཆོག has no duplicate, and we had already pushed to S3.

In [93]:
# Convert to DataFrame
df_1st = pd.DataFrame(list(first_links.items()), columns=['file_name', 'links'])
df_1st["data_path"] = "./data/གོང་ས་མཆོག/"
df_1st["audio_path"] = "./data/audio/གོང་ས་མཆོག/"
df_1st.head()

,file_name,links,data_path,audio_path
0,RFA_གོང་ས་མཆོག_1024,https://www.rfa.org/tibetan/dalai-lama/hhdl-in...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
1,RFA_གོང་ས་མཆོག_0,https://www.rfa.org/tibetan/sargyur/hhdl-us-it...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
2,RFA_གོང་ས་མཆོག_1025,https://www.rfa.org/tibetan/dalai-lama/mind-an...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
3,RFA_གོང་ས་མཆོག_1,https://www.rfa.org/tibetan/exile/tibetan-amer...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
4,RFA_གོང་ས་མཆོག_1026,https://www.rfa.org/tibetan/dalai-lama/26th-an...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/


In [94]:
# Convert to DataFrame
df_2 = pd.DataFrame(list(second_links.items()), columns=['file_name', 'links'])
df_2["data_path"] = "./data/བོད།/"
df_2["audio_path"] = "./data/audio/བོད།/"
df_2.head()

,file_name,links,data_path,audio_path
0,RFA_བོད།_1016,https://www.rfa.org/tibetan/tibet/cultural-rev...,./data/བོད།/,./data/audio/བོད།/
1,RFA_བོད།_0,https://www.rfa.org/tibetan/tibet/gangjong-she...,./data/བོད།/,./data/audio/བོད།/
2,RFA_བོད།_1036,https://www.rfa.org/tibetan/tibet/construction...,./data/བོད།/,./data/audio/བོད།/
3,RFA_བོད།_1,https://www.rfa.org/tibetan/sargyur/tibetan-re...,./data/བོད།/,./data/audio/བོད།/
4,RFA_བོད།_10,https://www.rfa.org/tibetan/sargyur/china-secu...,./data/བོད།/,./data/audio/བོད།/


In [95]:
# Convert to DataFrame
df_3 = pd.DataFrame(list(third_links.items()), columns=['file_name', 'links'])
df_3["data_path"] = "./data/བཙན་བྱོལ།/"
df_3["audio_path"] = "./data/audio/བཙན་བྱོལ།/"
df_3.head()

,file_name,links,data_path,audio_path
0,RFA_བཙན་བྱོལ།_0,https://www.rfa.org/tibetan/sargyur/tibetan-pa...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/
1,RFA_བཙན་བྱོལ།_1020,https://www.rfa.org/tibetan/sargyur/cta-delega...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/
2,RFA_བཙན་བྱོལ།_1011,https://www.rfa.org/tibetan/exile/sikyong-penp...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/
3,RFA_བཙན་བྱོལ།_1,https://www.rfa.org/tibetan/exile/tibetan-amer...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/
4,RFA_བཙན་བྱོལ།_1014,https://www.rfa.org/tibetan/sargyur/sikyong-sw...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/


In [97]:
# Convert to DataFrame
df_4 = pd.DataFrame(list(forth_links.items()), columns=['file_name', 'links'])
df_4["data_path"] = "./data/འཛམ་གླིང༌།/"
df_4["audio_path"] = "./data/audio/འཛམ་གླིང༌།/"
df_4.head()

,file_name,links,data_path,audio_path
0,RFA_འཛམ་གླིང༌།_0,https://www.rfa.org/tibetan/sargyur/sarah-coll...,./data/འཛམ་གླིང༌།/,./data/audio/འཛམ་གླིང༌།/
1,RFA_འཛམ་གླིང༌།_1026,https://www.rfa.org/tibetan/world/tibetan-dele...,./data/འཛམ་གླིང༌།/,./data/audio/འཛམ་གླིང༌།/
2,RFA_འཛམ་གླིང༌།_1010,https://www.rfa.org/tibetan/world/us-china-wil...,./data/འཛམ་གླིང༌།/,./data/audio/འཛམ་གླིང༌།/
3,RFA_འཛམ་གླིང༌།_1,https://www.rfa.org/tibetan/sargyur/tibetan-pa...,./data/འཛམ་གླིང༌།/,./data/audio/འཛམ་གླིང༌།/
4,RFA_འཛམ་གླིང༌།_1011,https://www.rfa.org/tibetan/society/tibetan-sw...,./data/འཛམ་གླིང༌།/,./data/audio/འཛམ་གླིང༌།/


In [99]:
# Convert to DataFrame
df_5 = pd.DataFrame(list(fif_links.items()), columns=['file_name', 'links'])
df_5["data_path"] = "./data/རྒྱ་དཀར་ནག/"
df_5["audio_path"] = "./data/audio/རྒྱ་དཀར་ནག/"
df_5.head()

,file_name,links,data_path,audio_path
0,RFA_རྒྱ་དཀར་ནག_0,https://www.rfa.org/tibetan/sargyur/china-lond...,./data/རྒྱ་དཀར་ནག/,./data/audio/རྒྱ་དཀར་ནག/
1,RFA_རྒྱ་དཀར་ནག_1026,https://www.rfa.org/tibetan/indiaandchina/taiw...,./data/རྒྱ་དཀར་ནག/,./data/audio/རྒྱ་དཀར་ནག/
2,RFA_རྒྱ་དཀར་ནག_1010,https://www.rfa.org/tibetan/indiaandchina/coro...,./data/རྒྱ་དཀར་ནག/,./data/audio/རྒྱ་དཀར་ནག/
3,RFA_རྒྱ་དཀར་ནག_1,https://www.rfa.org/tibetan/indiaandchina/chin...,./data/རྒྱ་དཀར་ནག/,./data/audio/རྒྱ་དཀར་ནག/
4,RFA_རྒྱ་དཀར་ནག_1011,https://www.rfa.org/tibetan/indiaandchina/chal...,./data/རྒྱ་དཀར་ནག/,./data/audio/རྒྱ་དཀར་ནག/


In [101]:
# Convert to DataFrame
df_6 = pd.DataFrame(list(six_links.items()), columns=['file_name', 'links'])
df_6["data_path"] = "./data/སྤྱི་ཚོགས།/"
df_6["audio_path"] = "./data/audio/སྤྱི་ཚོགས།/"
df_6.head()

,file_name,links,data_path,audio_path
0,RFA_སྤྱི་ཚོགས།_0,https://www.rfa.org/tibetan/society/geshe-trin...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
1,RFA_སྤྱི་ཚོགས།_1026,https://www.rfa.org/tibetan/society/sonam-jamy...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
2,RFA_སྤྱི་ཚོགས།_1010,https://www.rfa.org/tibetan/society/canada-rtw...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
3,RFA_སྤྱི་ཚོགས།_1,https://www.rfa.org/tibetan/society/tibetan-ar...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
4,RFA_སྤྱི་ཚོགས།_1011,https://www.rfa.org/tibetan/society/webinar-on...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/


## merge all

In [102]:
merged_df = pd.concat([df_1st, df_2, df_3, df_4, df_6, df_5])
merged_df.shape

(18425, 4)

In [103]:
clean_merged_df = merged_df.drop_duplicates(subset=['links'])
clean_merged_df.shape

(16386, 4)

In [104]:
clean_merged_df.sample(10)

,file_name,links,data_path,audio_path
3429,RFA_སྤྱི་ཚོགས།_589,https://www.rfa.org/tibetan/society/pema-tsete...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
524,RFA_གོང་ས་མཆོག_1457,https://www.rfa.org/tibetan/dalai-lama/his-hol...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
1707,RFA_འཛམ་གླིང༌།_2540,https://www.rfa.org/tibetan/world/taiwan-elect...,./data/འཛམ་གླིང༌།/,./data/audio/འཛམ་གླིང༌།/
1122,RFA_བོད།_2025,https://www.rfa.org/tibetan/tibet/one-killed-m...,./data/བོད།/,./data/audio/བོད།/
497,RFA_སྤྱི་ཚོགས།_1453,https://www.rfa.org/tibetan/society/cold-fever...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
1860,RFA_བཙན་བྱོལ།_2680,https://www.rfa.org/tibetan/exile/2016-machik-...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/
1479,RFA_སྤྱི་ཚོགས།_2318,https://www.rfa.org/tibetan/society/free-medic...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
617,RFA_བཙན་བྱོལ།_154,https://www.rfa.org/tibetan/exile/nedhon-lengm...,./data/བཙན་བྱོལ།/,./data/audio/བཙན་བྱོལ།/
2301,RFA_སྤྱི་ཚོགས།_3064,https://www.rfa.org/tibetan/society/tamding-ma...,./data/སྤྱི་ཚོགས།/,./data/audio/སྤྱི་ཚོགས།/
107,RFA_བོད།_1108,https://www.rfa.org/tibetan/tibet/bilingual-ed...,./data/བོད།/,./data/audio/བོད།/


In [105]:
file_name = "RFATibetan_data_map.csv"
clean_data_dir = './data/'
clean_merged_df.to_csv(clean_data_dir + file_name, index=False)

In [107]:
# Convert to DataFrame
df_7 = pd.DataFrame(list(sev_links.items()), columns=['file_name', 'links'])
df_7["data_path"] = "./data/གསར་འགྱུར།/"
df_7["audio_path"] = "./data/audio/གསར་འགྱུར།/"
df_7.head()

,file_name,links,data_path,audio_path
0,RFA_གསར་འགྱུར།_34,https://www.rfa.org/tibetan/sargyur/cta-monast...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
1,RFA_གསར་འགྱུར།_2,https://www.rfa.org/tibetan/sargyur/thinley-lh...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
2,RFA_གསར་འགྱུར།_3,https://www.rfa.org/tibetan/sargyur/richard-r-...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
3,RFA_གསར་འགྱུར།_4,https://www.rfa.org/tibetan/sargyur/tibetan-pl...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
4,RFA_གསར་འགྱུར།_5,https://www.rfa.org/tibetan/sargyur/tpie-austr...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/


In [108]:
merged_df = pd.concat([df_1st, df_2, df_3, df_4, df_6, df_5, df_7])
merged_df.shape

(47172, 4)

In [109]:
clean_merged_df = merged_df.drop_duplicates(subset=['links'])
clean_merged_df.shape

(43919, 4)

In [110]:
clean_merged_df.sample(10)

,file_name,links,data_path,audio_path
1513,RFA_བོད།_2347,https://www.rfa.org/tibetan/tibet/raise-the-Ch...,./data/བོད།/,./data/audio/བོད།/
17917,RFA_གསར་འགྱུར།_18891,https://www.rfa.org/tibetan/sargyur/government...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
4675,RFA_གསར་འགྱུར།_4675,https://www.rfa.org/tibetan/sargyur/eu-china-t...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
27720,RFA_གསར་འགྱུར།_28694,https://www.rfa.org/tibetan/sargyur/126973-200...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
5740,RFA_གསར་འགྱུར།_5740,https://www.rfa.org/tibetan/sargyur/dalai-lama...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
26396,RFA_གསར་འགྱུར།_27370,https://www.rfa.org/tibetan/sargyur/taiwan-pre...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
1504,RFA_གོང་ས་མཆོག_710,https://www.rfa.org/tibetan/dalai-lama/hhdl-ev...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
755,RFA_གོང་ས་མཆོག_1651,https://www.rfa.org/tibetan/dalai-lama/dalai-l...,./data/གོང་ས་མཆོག/,./data/audio/གོང་ས་མཆོག/
22343,RFA_གསར་འགྱུར།_23317,https://www.rfa.org/tibetan/sargyur/developmen...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/
21871,RFA_གསར་འགྱུར།_22845,https://www.rfa.org/tibetan/sargyur/world-soci...,./data/གསར་འགྱུར།/,./data/audio/གསར་འགྱུར།/


In [111]:
file_name = "RFATibetan_data_map_all.csv"
clean_data_dir = './data/'
clean_merged_df.to_csv(clean_data_dir + file_name, index=False)

In [112]:
# import pandas as pd
# import shutil
# import os


# # Define the clean data destination directory
# clean_data_dir = './data/clean_data/'

# # Create the clean data directory if it doesn't exist
# os.makedirs(clean_data_dir, exist_ok=True)

# # Iterate through the DataFrame rows
# for index, row in clean_merged_df.iterrows():
#     # Construct source and destination file paths
#     json_source = f"{row['data_path']}{row['file_name']}.json"
#     audio_source = f"{row['audio_path']}{row['file_name']}.mp3"
    
#     json_dest = f"{clean_data_dir}{row['file_name']}.json"
#     audio_dest = f"{clean_data_dir}{row['file_name']}.mp3"
    
#     # Copy JSON file
#     try:
#         shutil.copy2(json_source, json_dest)
#         print(f"Copied JSON: {json_source} -> {json_dest}")
#     except FileNotFoundError:
#         print(f"JSON file not found: {json_source}")
#     except Exception as e:
#         print(f"Error copying JSON {json_source}: {str(e)}")
    
#     # Copy MP3 file
#     try:
#         shutil.copy2(audio_source, audio_dest)
#         print(f"Copied MP3: {audio_source} -> {audio_dest}")
#     except FileNotFoundError:
#         print(f"MP3 file not found: {audio_source}")
#     except Exception as e:
#         print(f"Error copying MP3 {audio_source}: {str(e)}")

# print("File copying complete!")